# Solutions for pythonchallenge.com puzzles

Most are implemented in Python, but some are also solved in C++.

## [0. warming up](http://www.pythonchallenge.com/pc/def/0.html)
#### Problem:
This is a picture of 2**38.
#### Solution:
This is a simple intro. Just use pow() or 2**38.

In [59]:
2**38

274877906944

#### Answer:
274877906944

## [1. map](http://www.pythonchallenge.com/pc/def/map.html)

#### Problem:
Decode the messages.

#### Hints:
- The page title is 'What about making trans?'
- The page's image display a mapping:
    + K --> M
    + O --> Q
    + E --> G
    
#### Solution:
This is a basic Ceasar's cypher.

The hints suggests using string.maketrans().  This works, but it requires a translation table. It also seems to have issues in Python 3. string.maketrans was deprecated and then removed.  Strings are not bytes. They are Unicode.

I had considered using a dictionary LUT, but felt it makes reusability much harder. If the mapping K--> M changed to something else like K --> P, then a new table is required. One could write a function to generate the LUT or translation table for the required shift.   

Instead, string.ascii_lowercase and string.find() were used to get a character's index in the ascii string.  Then it becomes a matter of handling the wrapping. Positive/negative shifts and upper/lower case strings are handled.

#### Answer: ocr

In [60]:
%cd /home/ohlemacher/dev/pychallenge/1_think_twice
import think
think = reload(think)
msg_orig = \
        "g fmnc wms bgblr rpylqjyrc gr zw fylb. " + \
        "rfyrq ufyr amknsrcpq ypc dmp. bmgle gr gl zw fylb gq glcddgagclr" + \
        "ylb rfyr'q ufw rfgq rcvr gq qm jmle. sqgle qrpgle.kyicrpylq() gq pcamkkclbcb. lmu ynnjw ml rfc spj."
think.explore(msg_orig, 2)

/home/ohlemacher/dev/pychallenge/1_think_twice

Encrypted message:
g fmnc wms bgblr rpylqjyrc gr zw fylb. rfyrq ufyr amknsrcpq ypc dmp. bmgle gr gl zw fylb gq glcddgagclrylb rfyr'q ufw rfgq rcvr gq qm jmle. sqgle qrpgle.kyicrpylq() gq pcamkkclbcb. lmu ynnjw ml rfc spj.

Decrypted message:
i hope you didnt translate it by hand. thats what computers are for. doing it in by hand is inefficientand that's why this text is so long. using string.maketrans() is recommended. now apply on the url.


In [55]:
%cd /home/ohlemacher/dev/pychallenge/1_think_twice
import think
think = reload(think)
msg_orig = "map"
think.explore(msg_orig, 2)

/home/ohlemacher/dev/pychallenge/1_think_twice

Encrypted message:
map

Decrypted message:
ocr


In [58]:
print think.shift_msg('Pdeo skqhz xa qoabqh.', 4)
print think.shift_msg('This would be useful.', -4)

This would be useful.
Pdeo skqhz xa qoabqh.


## [2. ocr](http://www.pythonchallenge.com/pc/def/ocr.html)
#### Problem:
A picture of a book open to some unreadable text.

#### Hints:
- The image has a caption 'recognize the characters. maybe they are in the book, but MAYBE they are in the page' source.
- The source contains 'find rare characters in the mess below:'
- The mess is a large block of text '%%$@_$^__#)^)&!_+]!*@&^}....'

#### Solution:
Count the characters in the text block. The answer is found in the characters that are used only once in the order they appear.

A dictionary {[char: count]} was used to collect the counts, then dictionary was sorted: `sorted(dic.items(), key=operator.itemgetter(1))`. Fortunately, the built-in sorted() function is guaranteed to be stable. A sort is stable if it guarantees not to change the relative order of elements that compare equal.

#### Answer: equality